README
****************************************
USAGE - Python 2.7.x or 3.5.x
****************************************

1) from eip import PLC

2) test = PLC()

3) Set the IP Address: test.IPAddress = "192.168.1.10"

4) If your processor is in a slot other than 0, set that: test.ProcessorSlot = 1

4a) For Micro8xx controllers, use test.Micro800 = True

5) Do one of the following:

    Read a value:  value = test.Read("tagname")
    x: value = test.Read("MyTimer.ACC")
        print(value)
    *Reading a unique tag can be sped up by specifying the data type
        ahead of time: test.Read("MyDINT", datatype=196)
        
    Read an array: value = test.Read("tagname", length)
    ex: value = test.Read("MyDINT", 10)
        print(value(3))
    
    Read multiple tags at once: test.MultiRead("tag1", "tag2", "tag3")
    ex: values = test.MultiRead("MyDINT", "MyTimer.ACC", "AnotherDINT")
        print values

    You can also read multiple tags now by passing a list of tags to Read()
    ex: tags = ["MyDint", "AnotherDINT', "ThisREAL"]
        values = test.Read(tags)
        print(values)
    *MultiRead still works, but will probably go away eventually.

    Write a value: test.Write("tagname", value)
    ex: test.Write("MyDINT", 1337)
    *Writing a unique tag can be sped up by specifying the data type
        ahead of time: test.Wtite("MyDINT", 1337, datatype=196)

    Write an array of values:
    vals = [1234, 56789]
    test.Write("MyDint[0]", vals)

    Get the controller scoped tags
    ex: tags = test.GetTagList()
        (returns a structure of .TagName, .DataType and .Offset)

    Get all Ethernet I/P devices on the network test.Discover()
    ex:  stuff = test.Discover()
        print(stuff[1].IPAddress) # prints the first devices IP Address

    Get properties of a module in a particular slot
        thing = test.GetModuleProperties(0)
        print(thing.ProductName, thing.SerialNumber)
        (returns LGXDevice(), though not all members apply)

    Get or Set the PLC clock: test.GetPLCTime() or test.SetPLCTime()
    Note:  SetPLCTime() will use your computers clock to set the PLC

6) Close the connection with test.Close().  This should be called when you are
    completely done reading/writing, not with each exchange of data.

So a complete script to read a tag would look like this:

from eip import PLC
test = PLC()
test.IPAddress = "192.168.1.64"
value = test.Read("MyTimer.ACC")
print value
test.Close()


Another method is by using a with statement:

from eip import PLC
with PLC() as test:
    test.IPAddress = "192.168.1.64"
    print test.Read("MyTimer.ACC")




****************************************
NOTES
****************************************

Several things take place when making a connection to a PLC.

1) Standard TCP connection
2) Session Registration, PLC will respond with a Session Handle
3) Forward Open.  Using the Session Handle, this establishes
        the parameters for connection
4) Command (read/write/etc) and payload

The initial connections (TCP, Session Registration, Forward Open)
    only need to be made once, after that, we can exchange data
    until the connection is closed or lost.

A number events that take place on the command end, they should
    be invisible to the user, but it's good to know (and a reminder for me).
    The user can be reading/writing simple tags, UDT's arrays, bits of a word,
    bools out of atomic arrays, custom length strings, array reads that
    won't fit in a single reply and so on.  Each has to be handled in a
    different way.

    It basically happens like this:  Figure out if its a single tag to be
    read/written or an array.  Open the connection to the PLC (1,2,3 above),
    parse the tag name, add the command (read/write), send the data to the PLC,
    handle the reply

In order to perform a typical read/write, the data type needs to be known, or
    more specifically, the number of bytes the tag is.  It would be pretty
    annoying for the user to have to specify this each time, so to get around
    this, we have the InitialRead function.  This is called on any data exchange
    where the tag name has never been read/written before.  It uses a different
    CIP command than a typical read (0x52 vs 0x4C) which will respond with the
    data type.  This helps us in two ways:  First, the user won't need to specify
    the data type.  Second, we can keep track of this information so the next
    time the particular tag is called, we already know it's type so we can just
    do the normal read/write (0x4C/0x4D/0x4E).


changelog
04/21/21
  - Save tag list in TagList

03/05/21
  - Fixed crashing when reading/writing and connection is lost

02/16/21
  - Fixed requesting module properties only requesting slot 0

02/12/21
  - Fixed reading large boolean arrays

01/13/21
  - Reference data types by their hex value
  - Rename StructIdentifier > StringID

01/02/21
  - Added video link for unit testing demo on tests/README.md

01/01/21
  - Fix for reading BOOL arrays
  - Upped version to 0.7.8
  - Updated README

11/12/20
  - Fixed supplying data type when reading lists

11/11/20
  - Fixed returning only the first array value when reading lists

10/23/20
  - PR146
  - Fixes to continue python2 support

09/30/20
  - Rewrote parsing tag names
  - Fixed writing bits of a word
  - Upped version to 0.7.5

09/11/20
  - Updated tag list filter
  - Fixed timer example

09/10/20
  - Fixed getting struct raw data that was larger than a single packet

09/08/20
  - Fixed reading arrays larger than one packet
  - Fixed returning raw data when reading structs
  - Upped version to 0.7.4

09/04/20
  - Fixed import issue

09/02/20
  - Moved communications to its own module
  - Upped version to 0.7.3

09/01/20
  - Moved some code to lgx_device and lgx_tag
  - Removed unnecessary calls to _getUDT()
  - Upped version to 0.7.2

08/24/20
  - Fixed typo in lgx_device.py

07/15/20
  - Check connection sooner when reading lists
  - Updated README

07/14/20
  - Use muti service object to learn data types
  - Handle single leftover tag when getting multiple data types

07/13/20
  - Fixed when single tag is left over in batch read

07/09/20
  - Fixes for Micro800 strings

07/08/20
  - Fixed read issues after reorganization
  - Upped version to 0.6.11
  - Cleanup leftovers

07/07/20
  - Reorganized read and parsing replies
  - DWORD had incorrect pack/unpack format
  - Fixed getwrordcount returning a float under python3

05/13/20
  - Force little endian to support pylogix import in jython as Java uses big-endian. Issue #132 

05/13/20
  - Fixed PylogixTests.py imports

05/11/20
  - Default to LargeForwardOpen, fall back on ForwardOpen
  - Restructured some code, cleanup

05/05/20
  - Moved templates to .github hidden folder

05/04/20
  - Fix raw return for GetPLCTime
  - Fix printed return for example 21
  
04/23/20
  - jprogin's enhancements (read large lists, save UDT)
  - Fixes for making too many requests for templates already known
  - Upped version

4/01/20
  - Added Docs folder
  - Added Templates for Issues and Pull requests
  - Added Guides
  - Added UnitTesting doc
  - Upped version to 0.6.6
  - Updated README to reflect pypi install

03/29/20
  - Fixed routing to Micro800 controllers
  - Fixed GetDeviceProperties for Micro800
  - Fixed ForwardClose
  - Randomize TOConnectionID
  - Upped version to 0.6.5

03/24/20
  - Added the ability to route packets through a CLX rack, upped to 0.6.4

03/20/20
  - Impoved lost connections and error handling, upped version to 0.6.3

02/28/20
  - Fixed typo in lgxDevice.py
  - Upped to version 0.6.2

02/19/20
  - Changed connection to properly allow unconnected send for getting module properties
  - Added GetDeviceProperties to get properties of a device at the specified address
  - Added str() and repr() methods to LgxTag and LGXDevice
  - Upped version to 0.6.1

01/16/20
  - Updated readme

01/04/20
  - Upped version to 0.6.0, skipping the 0.5.x series
  - Fix for writing strings to Micro800

12/20/19
  - Upped version to 0.5.0
  - Removed deprecated MultiRead method since Read handles this
  - Implemented multi-write service
  - Just some renaming and cleanup
  - More cleanup and renaming

12/17/19
  - Added fragmented write for arrays larger than a packet

10/29/19
  - Moved examples from pylogix to root directory
  - Added comments about how data is returned
  - Added IP address and slot as parameters to PLC
  - Upped version to 0.4.12

10/02/19
  - setup.py imports version from __init__
  - update tests\readme with tox info

10/01/19
  - Added tox.ini for automated tests

10/01/19
  - Added unittest for clx, and compactlogix.
  - eip.py changes were fixed according to PEP8

09/30/19
  -Fixed format of str & repr in Response
  
09/27/19 - kodaman2
  - Added Repr, and str to Response class as per #77

09/27/19 - kodaman2
  - Fixed Multiple paths for status on GetProgramsList

09/27/19 - kodaman2
  - Modified to return correct return status, when nesting Response.Status to another Response object

09/25/19 - kodaman2
  - Path for GetProgramTagList to return LGXTag properly

09/25/19 - kodaman2
  - Fixed _writeTag to non existent tags exception error

09/25/19 - kodaman2
  - Updated README with old installation instructions

09/23/19 - kodaman2
  - Addressed GetProgramsList and GetProgramTagList issues

09/23/19 - ambersnow
  - Fixed GetTagList() AtributeError

09/18/19
  -Changed the case of Response to match the rest of the project, minimizing the impact of the new change

09/17/19
  -Rewrote all methods to return Response class, which includes tag_name, value and status
  -Updated README

08/26/19
  -Updated gitignore
  -Lowered default connection size to support older hardware/firmware by default

08/08/19
  -Fixed writing LREAL data types

07/18/19
  -Fixed imports, a few erros from restructure

07/09/19
  -cleanup
  -Restructured class a bit
  -Fixed reading tuples
  -Cleaned up imports
  
06/13/19
  -Fixed reading a list of a single item

05/31/19
  -Fixed connection parameters that cause Micro8xx to fail forwrad open

05/28/19
  -Fixed closing socket, suggested by Derriick

05/05/19
  -Pulled extneded forward open support from aansel-plenty
  -Did a little code cleaup
  -Fixed crash reading large chunks of data
  -Make sure connection size is correct
  -Added 50_Large_Forward_Open example

04/08/19
  -Return the raw bytes when a UDT is read, added example
  -Update install documentation
  -Allow passing date type when multi-reading for faster reading

03/13/19

context.py

def value(index):
  if index==0: return 0x6572276557
  if index==1: return 0x6f6e
  if index==2: return 0x676e61727473
  if index==3: return 0x737265
  if index==4: return 0x6f74
  if index==5: return 0x65766f6c
  if index==6: return 0x756f59
  if index==7: return 0x776f6e6b
  if index==8: return 0x656874
  if index==9: return 0x73656c7572
  if index==10: return 0x646e61
  if index==11: return 0x6f73
  if index==12: return 0x6f64
  if index==13: return 0x49
  if index==14: return 0x41
  if index==15: return 0x6c6c7566
  if index==16: return 0x74696d6d6f63
  if index==17: return 0x7327746e656d
  if index==18: return 0x74616877
  if index==19: return 0x6d2749
  if index==20: return 0x6b6e696874
  if index==21: return 0x676e69
  if index==22: return 0x666f
  if index==23: return 0x756f59
  if index==24: return 0x746e646c756f77
  if index==25: return 0x746567
  if index==26: return 0x73696874
  if index==27: return 0x6d6f7266
  if index==28: return 0x796e61
  if index==29: return 0x726568746f
  if index==30: return 0x797567
  if index==31: return 0x49
  if index==32: return 0x7473756a
  if index==33: return 0x616e6e6177
  if index==34: return 0x6c6c6574
  if index==35: return 0x756f79
  if index==36: return 0x776f68
  if index==37: return 0x6d2749
  if index==38: return 0x676e696c656566
  if index==39: return 0x6174746f47
  if index==40: return 0x656b616d
  if index==41: return 0x756f79
  if index==42: return 0x7265646e75
  if index==43: return 0x646e617473
  if index==44: return 0x726576654e
  if index==45: return 0x616e6e6f67
  if index==46: return 0x65766967
  if index==47: return 0x756f79
  if index==48: return 0x7075
  if index==49: return 0x726576654e
  if index==50: return 0x616e6e6f67
  if index==51: return 0x74656c
  if index==52: return 0x756f79
  if index==53: return 0x6e776f64
  if index==54: return 0x726576654e
  if index==55: return 0x616e6e6f67
  if index==56: return 0x6e7572
  if index==57: return 0x646e756f7261
  if index==58: return 0x646e61
  if index==59: return 0x747265736564
  if index==60: return 0x756f79
  if index==61: return 0x726576654e
  if index==62: return 0x616e6e6f67
  if index==63: return 0x656b616d
  if index==64: return 0x756f79
  if index==65: return 0x797263
  if index==66: return 0x726576654e
  if index==67: return 0x616e6e6f67
  if index==68: return 0x796173
  if index==69: return 0x657962646f6f67
  if index==70: return 0x726576654e
  if index==71: return 0x616e6e6f67
  if index==72: return 0x6c6c6574
  if index==73: return 0x61
  if index==74: return 0x65696c
  if index==75: return 0x646e61
  if index==76: return 0x74727568
  if index==77: return 0x756f79
  if index==78: return 0x6576276557
  if index==79: return 0x6e776f6e6b
  if index==80: return 0x68636165
  if index==81: return 0x726568746f
  if index==82: return 0x726f66
  if index==83: return 0x6f73
  if index==84: return 0x676e6f6c
  if index==85: return 0x72756f59
  if index==86: return 0x73277472616568
  if index==87: return 0x6e656562
  if index==88: return 0x676e69686361
  if index==89: return 0x747562
  if index==90: return 0x657227756f59
  if index==91: return 0x6f6f74
  if index==92: return 0x796873
  if index==93: return 0x6f74
  if index==94: return 0x796173
  if index==95: return 0x7469
  if index==96: return 0x656469736e49
  if index==97: return 0x6577
  if index==98: return 0x68746f62
  if index==99: return 0x776f6e6b
  if index==100: return 0x732774616877
  if index==101: return 0x6e656562
  if index==102: return 0x676e696f67
  if index==103: return 0x6e6f
  if index==104: return 0x6557
  if index==105: return 0x776f6e6b
  if index==106: return 0x656874
  if index==107: return 0x656d6167
  if index==108: return 0x646e61
  if index==109: return 0x6572276577
  if index==110: return 0x616e6e6f67
  if index==111: return 0x79616c70
  if index==112: return 0x7469
  if index==113: return 0x646e41
  if index==114: return 0x6669
  if index==115: return 0x756f79
  if index==116: return 0x6b7361
  if index==117: return 0x656d
  if index==118: return 0x776f68
  if index==119: return 0x6d2749
  if index==120: return 0x676e696c656566
  if index==121: return 0x74276e6f44
  if index==122: return 0x6c6c6574
  if index==123: return 0x656d
  if index==124: return 0x657227756f79
  if index==125: return 0x6f6f74
  if index==126: return 0x646e696c62
  if index==127: return 0x6f74
  if index==128: return 0x656573
  if index==129: return 0x726576654e
  if index==130: return 0x616e6e6f67
  if index==131: return 0x65766967
  if index==132: return 0x756f79
  if index==133: return 0x7075
  if index==134: return 0x726576654e
  if index==135: return 0x616e6e6f67
  if index==136: return 0x74656c
  if index==137: return 0x756f79
  if index==138: return 0x6e776f64
  if index==139: return 0x726576654e
  if index==140: return 0x616e6e6f67
  if index==141: return 0x6e7572
  if index==142: return 0x646e756f7261
  if index==143: return 0x646e61
  if index==144: return 0x747265736564
  if index==145: return 0x756f79
  if index==146: return 0x726576654e
  if index==147: return 0x616e6e6f67
  if index==148: return 0x656b616d
  if index==149: return 0x756f79
  if index==150: return 0x797263
  if index==151: return 0x726576654e
  if index==152: return 0x616e6e6f67
  if index==153: return 0x796173
  if index==154: return 0x657962646f6f67
  if index==155: return 0x726576654e
  if index==156: return 0xa680e2616e6e6f67
